actor_library/src/access_control/access_control.rs

```diff
@@ -0,0 +1,236 @@
+use cid::Cid;
+use fil_actors_runtime::{ActorError, AsActorError};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_hamt::BytesKey;
+use fvm_shared::{
+    address::Address,
+    error::ExitCode,
+    event::{Entry, Flags},
+    IPLD_RAW,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::context::Context;
+use crate::utils::map;
+
+const ROLE_ADMIN_CHANGED_EVENT_KEY: &str = "role_admin_changed_event";
+const ROLE_GRANTED_EVENT_KEY: &str = "role_granted_event";
+const ROLE_REVOKED_EVENT_KEY: &str = "role_revoked_event";
+
+#[derive(Serialize, Deserialize)]
+pub struct RoleAdminChangedEvent {
+    pub role: Role,
+    pub previous_admin_role: Role,
+    pub new_admin_role: Role,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RoleGrantedEvent {
+    pub role: Role,
+    pub account: Address,
+    pub sender: Address,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RoleRevokedEvent {
+    pub role: Role,
+    pub account: Address,
+    pub sender: Address,
+}
+
+pub type Role = u32;
+
+pub trait AccessControl {
+    // Modifier
+    fn only_role<BS: Blockstore>(&self, store: &BS, role: Role);
+
+    fn has_role<BS: Blockstore>(&self, store: &BS, role: Role, account: Address) -> bool;
+
+    fn get_role_admin<BS: Blockstore>(&self, store: &BS, role: Role) -> Role;
+
+    fn grant_role<BS: Blockstore>(&mut self, store: &BS, role: Role, account: Address);
+
+    fn revoke_role<BS: Blockstore>(&mut self, store: &BS, role: Role, account: Address);
+
+    fn renounce_role<BS: Blockstore>(&mut self, store: &BS, role: Role, account: Address);
+}
+
+pub trait AccessControlPrivate {
+    fn _check_role<BS: Blockstore>(&self, store: &BS, role: Role);
+
+    fn _check_role_and_account<BS: Blockstore>(&self, store: &BS, role: Role, account: Address);
+
+    fn _set_role_admin<BS: Blockstore>(&mut self, store: &BS, role: Role, admin_role: Role);
+
+    fn _grant_role<BS: Blockstore>(&mut self, store: &BS, role: Role, account: Address);
+
+    fn _revoke_role<BS: Blockstore>(&mut self, store: &BS, role: Role, account: Address);
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct DefaultAccessControl {
+    // Mapping(Role+Address -> _)
+    roles: Cid,
+
+    // Mapping(Role -> AdminRole)
+    admin_roles: Cid,
+}
+
+impl Context for DefaultAccessControl {}
+
+impl DefaultAccessControl {
+    pub fn new<BS: Blockstore>(store: &BS) -> Result<Self, ActorError> {
+        let empty_role_map = map::make_empty_map::<_, ()>(store).flush().context_code(
+            ExitCode::USR_ILLEGAL_STATE,
+            "Failed to create empty role map",
+        )?;
+
+        let empty_admin_role_map = map::make_empty_map::<_, Role>(store).flush().context_code(
+            ExitCode::USR_ILLEGAL_STATE,
+            "Failed to create empty admin role map",
+        )?;
+
+        Ok(Self {
+            roles: empty_role_map,
+            admin_roles: empty_admin_role_map,
+        })
+    }
```

Comment:
Add methods from AccessControlPrivate trait here.
```suggestion
    }

    pub fn _check_role<BS: Blockstore>(&self, store: &BS, role: Role) {
        Self::_check_role_and_account(self, store, role, DefaultAccessControl::msg_sender());
    }

    pub fn _check_role_and_account<BS: Blockstore>(&self, store: &BS, role: Role, account: Address) {
        assert_eq!(
            Self::has_role(self, store, role, account),
            true,
            "Access Control: account {} is missing role {}",
            account,
            role
        );
    }

    etc.
```

Reply:
Methods added [here](https://github.com/Ethernal-Tech/filecoin-actor-library/pull/5/commits/dcd00ad0e72ae3cb2412881b9b4a8a415dcf9860#diff-324b0e57c0074407a39059591d361b952bc9f0f06b445d4376767d222ba55ef0R103)